Add bipartite adjacency maps

vasalf opened the pull request.

snowleopard commented on src/Algebra/Graph/Bipartite/AdjacencyMap.hs:
```diff
@@ -0,0 +1,24 @@
+} deriving Eq
+
+consistent :: (Ord a, Ord b) => AdjacencyMap a b -> Bool
+consistent (BAM lr rl) = oneWayConstistent lr rl && oneWayConstistent rl lr
```
The current invariant seems too complicated. Couldn't we just do this?

```haskell
consistent (BAM lr rl) = internalEdgeList lr == sort (map swap $ internalList rl)
```

vasalf replied:
@snowleopard, good point, it looks much easier to understand this way!

snowleopard commented on src/Algebra/Graph/Bipartite/AdjacencyMap.hs:
```diff
@@ -0,0 +1,24 @@
+internalEdgeList :: Map.Map a (Set.Set b) -> [(a, b)]
+internalEdgeList lr = [ (u, v) | (u, vs) <- Map.toAscList lr, v <- Set.toAscList vs ]
+
+hasEdge :: (Ord a, Ord b) => Map.Map a (Set.Set b) -> a -> b -> Bool
```
It would be best to make this function similar to the API of other modules:

```haskell
hasEdge :: (Ord a, Ord b) => a -> b -> AdjacencyMap a b -> Bool
```

vasalf replied:
@snowleopard It was first designed to be used in `consistent`. Now that we don't need it to check consistency, I will change the signature.

snowleopard commented on src/Algebra/Graph/Bipartite/AdjacencyMap.hs:
```diff
@@ -0,0 +1,165 @@
+consistent :: (Ord a, Ord b) => AdjacencyMap a b -> Bool
+consistent (BAM lr rl) = internalEdgeList lr == sort (map swap $ internalEdgeList rl)
+
+-- | The list of edges of a bipartite adjacency map
```
The convention in the library is to not give Haddock comments to functions that are not exported -- it makes it easier to immediately see whether a function is exported or not, so please change `-- |` to just `--`.

snowleopard replied:
Same for `addReverseEdges`.

snowleopard commented on src/Algebra/Graph/Bipartite/AdjacencyMap.hs:
```diff
@@ -0,0 +1,165 @@
+import qualified Data.Set        as Set
+
+-- | The 'Bipartite.AdjacencyMap' data type represents an __undirected__
+-- bipartite graph by two maps of vertices into its adjacency sets. The two
```
into its -> into their

snowleopard commented on src/Algebra/Graph/Bipartite/AdjacencyMap.hs:
```diff
@@ -0,0 +1,165 @@
+    -- memory
+    --
+    -- @
+    -- leftToRight ('fromGraph' 'Algebra.Graph.empty')                                          == Map.'Map.empty'
```
I think it may be better to follow the naming convention in the rest of the library, where such functions are typically named `adjacencyMap`. In this case we could call them `leftAdjacencyMap` and `rightAdjacencyMap`.

snowleopard commented on src/Algebra/Graph/Bipartite/AdjacencyMap.hs:
```diff
@@ -0,0 +1,165 @@
+-- | Check that the internal graph representation is consistent, i.e. that all
+-- edges that are present in the 'leftToRight' map are present in the
+-- 'rightToLeft' map.
+-- Complexity: /O(m log m)/
```
Let's drop complexity here. This function is not supposed to be performance-critical, so I'd say we don't want to promise any performance guarantees.

snowleopard commented on src/Algebra/Graph/Bipartite/AdjacencyMap.hs:
```diff
@@ -0,0 +1,165 @@
+-- consistent ('fromGraph' ('Algebra.Graph.vertex' x)) == True
+-- consistent ('fromGraph' ('Algebra.Graph.edge' x y)) == True
+-- consistent ('fromGraph' ('Algebra.Graph.edges' xs)) == True
+-- consistent ('fromGraph' g)          == True
```
This property covers all of the properties above, so I suggest to delete them.

vasalf replied:
I replaced them with the brand new primitives.

snowleopard commented on src/Algebra/Graph/Bipartite/AdjacencyMap.hs:
```diff
@@ -0,0 +1,165 @@
+-- 'leftToRight' (fromAdjacencyMap ('Algebra.Graph.AdjacencyMap.edge' (Left 1) (Left 2)))                       == Map.'Map.empty'
+-- 'rightToLeft' (fromAdjacencyMap ('Algebra.Graph.AdjacencyMap.edges' [(Left 1, Right 1), (Left 1, Right 2)])) == Map.'Map.fromAscList' [(1, Set.'Set.singleton' 1), (2, Set.'Set.singleton' 1)]
+-- @
+fromAdjacencyMap :: (Ord a, Ord b) => AM.AdjacencyMap (Either a b) -> AdjacencyMap a b
```
After checking other modules, I think this should be named `toBipartite` for consistency.

snowleopard commented on src/Algebra/Graph/Bipartite/AdjacencyMap.hs:
```diff
@@ -0,0 +1,165 @@
+-- toAdjacencyMap (fromGraph ('Algebra.Graph.vertex' $ Left 1))         == 'Algebra.Graph.AdjacencyMap.vertex' 1
+-- toAdjacencyMap (fromGraph ('Algebra.Graph.edge' (Left 1) (Right 2))) == 'Algebra.Graph.AdjacencyMap.edges' [(Left 1, Right 2), (Right 2, Left 1)]
+-- @
+toAdjacencyMap :: (Ord a, Ord b) => AdjacencyMap a b -> AM.AdjacencyMap (Either a b)
```
After checking other modules, I think this should be named `fromBipartite` for consistency.

snowleopard commented on src/Algebra/Graph/Bipartite/AdjacencyMap.hs:
```diff
@@ -0,0 +1,165 @@
+data AdjacencyMap a b = BAM {
+    -- | The /adjacency map/ of the left part of the graph: each vertex is
+    -- associated with a set of its neighbours. Complexity: /O(1)/ time and
+    -- memory
```
Missing dot (.) at the end of a sentence. Same issue in a couple of other places.

snowleopard commented on src/Algebra/Graph/Bipartite/AdjacencyMap.hs:
```diff
@@ -0,0 +1,165 @@
+addReverseEdges :: (Ord a, Ord b) => AM.AdjacencyMap (Either a b) -> AM.AdjacencyMap (Either a b)
+addReverseEdges m = AM.overlay m $ AM.edges [ (v, u) | (u, v) <- AM.edgeList m ]
+
+-- | Constructs a bipartite 'Bipartite.AdjacencyMap' from
```
I think the Haddock link is broken and should actually be `'AdjacencyMap'`? Please check generated documentation.

vasalf replied:
Yes, it was broken. Fixed now.

snowleopard commented on src/Algebra/Graph/Bipartite/AdjacencyMap.hs:
```diff
@@ -0,0 +1,165 @@
+addReverseEdges m = AM.overlay m $ AM.edges [ (v, u) | (u, v) <- AM.edgeList m ]
+
+-- | Constructs a bipartite 'Bipartite.AdjacencyMap' from
+-- 'Algebra.Graph.AdjacencyMap' with given part identifiers, adding all needed
```
Here please use double quotes to link to the module instead (otherwise it may be confusing to see two equally named links).

snowleopard commented on src/Algebra/Graph/Bipartite/AdjacencyMap.hs:
```diff
@@ -0,0 +1,165 @@
+    --
+    -- @
+    -- leftToRight ('fromGraph' 'Algebra.Graph.empty')                                          == Map.'Map.empty'
+    -- leftToRight ('fromGraph' ('Algebra.Graph.vertex' (Left 1)))                              == Map.'Map.singleton' 1 Set.'Set.empty'
```
The examples based on `fromGraph` conversions are quite difficult to read. Could you add basic graph construction primitives to this PR? I suggest adding at least `empty`, `leftVertex`, `rightVertex`, `edge` and `edges`. Then all examples in this module would be much nicer.

vasalf replied:
There are all of them now.

snowleopard commented:
@vasalf Many thanks for the PR! I've left a bunch of comments, please have a look.

nobrakal commented on src/Algebra/Graph/Bipartite/AdjacencyMap.hs:
```diff
@@ -0,0 +1,431 @@
+-- edgeCount . 'edges'    == 'length' . 'Data.List.nub'
+-- @
+edgeCount :: AdjacencyMap a b -> Int
+edgeCount = Map.foldr (+) 0 . fmap Set.size . leftAdjacencyMap
```
I am not sure that deforestation is implemented for `Map`... Maybe `Map.foldr ((+) . Set.size) 0` will be more efficient

vasalf replied:
Well, they claim in the docs that each of those `Map.*` functions work in linear time, so deforestation **is** implemented for strict maps. But still `fmap` has some overhead over `(.)`. In the following example `sumPlusOne'` works 1.5 times faster.

```Haskell
module Main where

import qualified Data.Map.Strict as Map

inputMap :: Map.Map Int Int
inputMap = Map.fromAscList [ (i, i) | i <- [1..1000000] ]

sumPlusOne :: Map.Map Int Int -> Int
sumPlusOne = Map.foldr (+) 0 . fmap ((+) 1)

sumPlusOne' :: Map.Map Int Int -> Int
sumPlusOne' = Map.foldr ((+) . ((+) 1)) 0

main :: IO ()
-- main = putStrLn $ show $ sumPlusOne inputMap
main = putStrLn $ show $ sumPlusOne' inputMap
```

I will change this to `Map.foldr ((+) . Set.size) 0`.

snowleopard replied:
Thanks for benchmarking, interesting result!

snowleopard commented on src/Algebra/Graph/Bipartite/AdjacencyMap.hs:
```diff
@@ -0,0 +1,431 @@
+-- the motivation behind the library, the underlying theory, and
+-- implementation details.
+--
+-- This module defines the 'Bipartite.AdjacencyMap' for bipartite graphs data
```
This Haddock link is most likely broken.

snowleopard commented on src/Algebra/Graph/Bipartite/AdjacencyMap.hs:
```diff
@@ -0,0 +1,431 @@
+    -- leftAdjacencyMap ('leftVertex' 1)           == Map.'Map.singleton' 1 Set.'Set.empty'
+    -- leftAdjacencyMap ('rightVertex' 1)          == Map.'Map.empty'
+    -- leftAdjacencyMap ('edge' 1 2)               == Map.'Map.singleton' 1 (Set.'Set.singleton' 2)
+    -- leftAdjacencyMap ('edge' 1 1)               == Map.'Map.singleton' 1 (Set.'Set.singleton' 1)
```
I suggest to first give the example `edge 1 1`, and then follow it with `edge 1 'a'` to demonstrate that the types of vertices in the two parts can be different.

snowleopard replied:
Same in `rightAdjacencyMap`.

snowleopard commented on src/Algebra/Graph/Bipartite/AdjacencyMap.hs:
```diff
@@ -0,0 +1,431 @@
+
+-- The list of edges of a bipartite adjacency map.
+--
+-- /Note: this function is for internal use only./
```
This comment is redundant: all non-exported functions are internal (and one can immediately see that this function is not exported, because it doesn't have a Haddock comment, by convention).

vasalf replied:
Whoops, forgot about it.

snowleopard commented on src/Algebra/Graph/Bipartite/AdjacencyMap.hs:
```diff
@@ -0,0 +1,431 @@
+-- hasEdge x y 'empty'                                == False
+-- hasEdge x y ('edge' x y)                           == True
+-- hasEdge 2 3 ('edge' 1 2)                           == False
+-- hasEdge x y ('overlay' g ('edge' x y))               == True
```
Let's use `x`-`z` in examples (instead of `g`).

snowleopard commented on src/Algebra/Graph/Bipartite/AdjacencyMap.hs:
```diff
@@ -0,0 +1,431 @@
+-- hasEdge 1 2 ('fromGraph' ('Algebra.Graph.edge' (Left 1) (Left 2))) == False
+-- @
+hasEdge :: (Ord a, Ord b) => a -> b -> AdjacencyMap a b -> Bool
+hasEdge u v (BAM m _) = ((Set.member v) <$> (u `Map.lookup` m)) == Just True
```
Redundant parentheses in `(Set.member v)`.

snowleopard commented on src/Algebra/Graph/Bipartite/AdjacencyMap.hs:
```diff
@@ -0,0 +1,431 @@
+hasRightVertex :: Ord b => b -> AdjacencyMap a b -> Bool
+hasRightVertex y (BAM _ rl) = y `Map.member` rl
+
+-- | Check if a graph contains a given vertex in the given part.
```
Can be shortened to just "Check if a graph contains a given vertex".

snowleopard commented on src/Algebra/Graph/Bipartite/AdjacencyMap.hs:
```diff
@@ -0,0 +1,431 @@
+--
+-- @
+-- leftVertexCount 'empty'         == 0
+-- leftVertexCount 'leftVertex' x  == 1
```
I don't think this compiles -- there are missing parentheses. Same below and in `rightVertexCount` examples.

snowleopard replied:
And in `vertexCount`.

snowleopard commented on src/Algebra/Graph/Bipartite/AdjacencyMap.hs:
```diff
@@ -0,0 +1,431 @@
+-- vertexCount 'empty'         == 0
+-- vertexCount 'leftVertex' x  == 1
+-- vertexCount 'rightVertex' x == 1
+-- vertexCount g             == 'leftVertexCount' g + 'rightVertexCount' g
```
`g` => `x`

snowleopard commented on src/Algebra/Graph/Bipartite/AdjacencyMap.hs:
```diff
@@ -0,0 +1,431 @@
+edgeCount = Map.foldr (+) 0 . fmap Set.size . leftAdjacencyMap
+
+-- Adds all edges needed to make the graph undirected.
+-- Complexity: /O(m log n)/.
```
Let's use `log(n)` consistently.

snowleopard commented on src/Algebra/Graph/Bipartite/AdjacencyMap.hs:
```diff
@@ -0,0 +1,431 @@
+-- Complexity: /O(m log n)/.
+--
+-- /Note: this function is for internal use only./
+addReverseEdges :: (Ord a, Ord b) => AM.AdjacencyMap (Either a b) -> AM.AdjacencyMap (Either a b)
```
I think this function is called `symmetricClosure` in other modules.

snowleopard replied:
Perhaps, you could even use `symmetricClosure` directly?

vasalf replied:
Yup, I just didn't notice it when I tried to search for something like this.

snowleopard commented on src/Algebra/Graph/Bipartite/AdjacencyMap.hs:
```diff
@@ -0,0 +1,431 @@
+    -- rightAdjacencyMap ('edges' [(1, 1), (1, 2)]) == Map.'Map.fromAscList' [(1, Set.'Set.singleton' 1), (2, Set.'Set.singleton' 1)]
+    -- @
+    rightAdjacencyMap :: Map.Map b (Set.Set a)
+} deriving (Eq, Show, Generic)
```
You shouldn't use derived `Show` instance here. Please have a look how `Show` is implemented for `AdjacencyMap` and try to follow it here.

snowleopard replied:
Also, using derived `Eq` is probably suboptimal: we can implement `Eq` by comparing only one of the two maps.

snowleopard replied:
Oh wait, my second comment is probably wrong :)

vasalf replied:
We can compare only `leftAdjacencyMap`s and `keysSet`s of the `rigthAdjacencyMap`s, this would be a little bit faster.

snowleopard commented on src/Algebra/Graph/Bipartite/AdjacencyMap.hs:
```diff
@@ -0,0 +1,431 @@
+fromGraph :: (Ord a, Ord b) => G.Graph (Either a b) -> AdjacencyMap a b
+fromGraph = toBipartite . (G.foldg AM.empty AM.vertex AM.overlay AM.connect)
+
+-- | Constructs the /empty graph/.
```
Please reorder all function definitions in this module according to the export list. This helps navigating it.

snowleopard replied:
Also, let's use "Construct" instead of "Constructs" throughout the module.

snowleopard commented on src/Algebra/Graph/Bipartite/AdjacencyMap.hs:
```diff
@@ -0,0 +1,431 @@
+-- Complexity: /O(m log n)/.
+--
+-- @
+-- 'leftAdjacencyMap' (fromGraph ('Algebra.Graph.empty'))                                         == Map.'Map.empty'
```
I think these examples can be simplified as follows:

```haskell
fromGraph (Algebra.Graph.'Algebra.Graph.empty') == 'empty'
fromGraph (Algebra.Graph.'Algebra.Graph.edge' (Left 1) (Right 2))) == 'edge' 1 2
...
```

snowleopard commented on src/Algebra/Graph/Bipartite/AdjacencyMap.hs:
```diff
@@ -0,0 +1,431 @@
+--
+-- @
+-- 'toBipartite ('Algebra.Graph.AdjacencyMap.empty')                                                           == 'empty'
+-- 'leftAdjacencyMap' (toBipartite ('Algebra.Graph.AdjacencyMap.edge' (Left 1) (Right 2)))                       == Map.'Map.singleton' 1 (Set.'Set.singleton' 2)
```
See this comment for how to simplify such examples: https://github.com/snowleopard/alga/pull/207/files#r295996443

snowleopard commented on src/Algebra/Graph/Bipartite/AdjacencyMap.hs:
```diff
@@ -0,0 +1,431 @@
+-- @
+-- 'leftAdjacencyMap' (leftVertex x)  == Map.'Map.singleton' x Set.'Set.empty'
+-- 'rightAdjacencyMap' (leftVertex x) == Map.'Map.empty'
+-- 'hasEdge' x y (leftVertex x)       == False
```
`leftVertex x` => `leftVertex z` for greater generality.

snowleopard commented on src/Algebra/Graph/Bipartite/AdjacencyMap.hs:
```diff
@@ -0,0 +1,431 @@
+-- 'leftAdjacencyMap' (edge x y)  == Map.'Map.singleton' x (Set.'Set.singleton' y)
+-- 'rightAdjacencyMap' (edge x y) == Map.'Map.singleton' y (Set.'Set.singleton' x)
+-- 'hasEdge' x y (edge x y)       == True
+-- 'hasEdge' y x (edge x y)       == (x == y)
```
I find this example a bit difficult to understand. Perhaps, it's worth splitting it into two concrete ones, say `x=y=1`, and `x=1` and `y=2`.

snowleopard commented on src/Algebra/Graph/Bipartite/AdjacencyMap.hs:
```diff
@@ -0,0 +1,431 @@
+overlay :: (Ord a, Ord b) => AdjacencyMap a b -> AdjacencyMap a b -> AdjacencyMap a b
+overlay (BAM lr1 rl1) (BAM lr2 rl2) = BAM (Map.unionWith Set.union lr1 lr2) (Map.unionWith Set.union rl1 rl2)
+
+-- | /Connect/ two bipartite graphs, not adding the inappropriate edges. This
```
Make it clearer: "not adding the inappropriate edges" => "not adding edges between vertices in the same part".

snowleopard commented on src/Algebra/Graph/Bipartite/AdjacencyMap.hs:
```diff
@@ -0,0 +1,431 @@
+
+-- | /Connect/ two bipartite graphs, not adding the inappropriate edges. This
+-- is an associative operation with the identity 'empty', which distributes
+-- over 'overlay' and obeys the decomposition axion.
```
"axion" => "axiom"

snowleopard commented on src/Algebra/Graph/Bipartite/AdjacencyMap.hs:
```diff
@@ -0,0 +1,431 @@
+overlay (BAM lr1 rl1) (BAM lr2 rl2) = BAM (Map.unionWith Set.union lr1 lr2) (Map.unionWith Set.union rl1 rl2)
+
+-- | /Connect/ two bipartite graphs, not adding the inappropriate edges. This
+-- is an associative operation with the identity 'empty', which distributes
```
I guess it's also commutative.

snowleopard commented on src/Algebra/Graph/Bipartite/AdjacencyMap.hs:
```diff
@@ -0,0 +1,431 @@
+-- number of vertices in the arguments: /m = O(m1 + m2 + l1 * r2 + l2 * r1)/.
+--
+-- @
+-- 'isEmpty'     (connect x y) == 'isEmpty'   x   && 'isEmpty'   y
```
Add a couple of concrete examples, as this function is quite tricky.

snowleopard commented on src/Algebra/Graph/Bipartite/AdjacencyMap.hs:
```diff
@@ -0,0 +1,431 @@
+connect :: (Ord a, Ord b) => AdjacencyMap a b -> AdjacencyMap a b -> AdjacencyMap a b
+connect (BAM lr1 rl1) (BAM lr2 rl2) = BAM (Map.unionsWith Set.union [ lr1, lr2,
+                                                                      Map.fromSet (const $ Map.keysSet rl2) (Map.keysSet lr1),
+                                                                      Map.fromSet (const $ Map.keysSet rl1) (Map.keysSet lr2)]) $
```
Please try not to exceed 80 characters per line, if possible. Long lines are difficult to review.

snowleopard commented on src/Algebra/Graph/Bipartite/AdjacencyMap.hs:
```diff
@@ -0,0 +1,431 @@
+-- 'hasLeftVertex' x . ('flip' ('const' [])) vertices == 'elem' x
+-- 'hasRightVertex' x . 'const' [] vertices         == 'elem' x
+-- @
+vertices :: (Ord a, Ord b) => [a] -> [b] -> AdjacencyMap a b
```
I think it may be better to follow the approach taken with `vertex`: create `leftVertices`, `rightVertices` and let `vertices` take a list of `Either`s as input. This will make the API closer to non-bipartite graphs.

vasalf replied:
OK, but maybe we would better rename the current version and keep it? Something like `fromPartLists`?
It seems ridiculus to do `overlay (leftVertices ls) (rightVertices rs)` or even `vertices ((Left <$> ls) ++ (Right <$> rs))` in case one needs it. And I needed it in some examples.

snowleopard replied:
OK, I agree, let's keep this function as is. Don't worry about adding others for now.

snowleopard commented on test/Algebra/Graph/Test/Bipartite/AdjacencyMap.hs:
```diff
@@ -0,0 +1,433 @@
+import Algebra.Graph.Test
+
+import qualified Algebra.Graph                       as G
+import qualified Algebra.Graph.AdjacencyMap          as AM
```
Whitespace.

adithyaov commented on test/Main.hs:
```diff
@@ -26,16 +27,17 @@ main :: IO ()
+    go "NonEmptyGraph"         testNonEmptyGraph
+    go "Relation"              testRelation
+    go "SymmetricRelation"     testSymmetricRelation
+    go "BipartiteAdjacencyMap" testBipartiteAdjacencyMap
```
Maybe put this after line 32 following the alphabetical order?

adithyaov commented on src/Algebra/Graph/Bipartite/AdjacencyMap.hs:
```diff
@@ -0,0 +1,433 @@
+-- the motivation behind the library, the underlying theory, and
+-- implementation details.
+--
+-- This module defines the 'AdjacencyMap' for bipartite graphs data type and
```
I think you meant `This module defines the 'AdjacencyMap' data type for bipartite graphs and`

snowleopard commented on src/Algebra/Graph/Bipartite/AdjacencyMap.hs:
```diff
@@ -0,0 +1,666 @@
+-- swap . 'vertices'   == flip 'vertices'
+-- swap ('edge' 1 "a") == 'edge' "a" 1
+-- swap . 'edges'      == 'edges' . map Data.Tuple.'Data.Tuple.swap'
+-- @
```
One of the nicest properties is missing: `swap . swap == id` :)

snowleopard commented on src/Algebra/Graph/Bipartite/AdjacencyMap.hs:
```diff
@@ -0,0 +1,666 @@
+-- consistent ('edge' x y)       == True
+-- consistent ('edges' xs)       == True
+-- consistent ('fromGraph' g)    == True
+-- consistent ('toBipartite' am) == True
```
Check consistency of `swap`.

snowleopard commented on src/Algebra/Graph/Bipartite/AdjacencyMap.hs:
```diff
@@ -0,0 +1,668 @@
+When specifying the time and memory complexity of graph algorithms, /n/ and /m/
+will denote the number of vertices and edges in the graph, respectively. In
+addition, /l/ and /r/ will denote the number of vertices in the left and in the
+right part of graph, respectively.
```
Just a note: I wonder if by adding `swap` into the algebra we could get some new interesting axioms and theorems. Something to think about!

snowleopard commented on src/Algebra/Graph/Bipartite/AdjacencyMap.hs:
```diff
@@ -0,0 +1,668 @@
+    -- rightAdjacencyMap 'empty'                    == Map.'Map.empty'
+    -- rightAdjacencyMap ('leftVertex' 1)           == Map.'Map.empty'
+    -- rightAdjacencyMap ('rightVertex' 1)          == Map.'Map.singleton' 2 Set.'Set.empty'
+    -- rightAdjacencyMap ('edge' 1 1)               == Map.'Map.singleton' 1 (Set.'Set.singleton' "a")
```
The above two lines seem wrong. Are they in the testsuite?

vasalf replied:
Whoops, this seems to be a copy-paste error. Yes, now they are.

snowleopard commented on src/Algebra/Graph/Bipartite/AdjacencyMap.hs:
```diff
@@ -0,0 +1,668 @@
+    negate      = id
+
+instance (Ord a, Ord b) => Eq (AdjacencyMap a b) where
+    (BAM lr1 rl1) == (BAM lr2 rl2) = (lr1 == lr2) && (Map.keysSet rl1 == Map.keysSet rl2)
```
This is a tricky definition. It would be nice to test that it actually produces the same result as the trivial implementation based on comparing both maps.

snowleopard commented on src/Algebra/Graph/Bipartite/AdjacencyMap.hs:
```diff
@@ -0,0 +1,668 @@
+-- vertices [] []                               == 'empty'
+-- vertices [x] []                              == 'leftVertex' x
+-- vertices [] [x]                              == 'rightVertex' x
+-- 'hasLeftVertex' x . ('flip' ('const' [])) vertices == 'elem' x
```
This and the next examples seem too complex, I suggest simplifying them by dropping point-free notation:

```haskell
hasLeftVertex  x (vertices ys zs) == elem x ys
hasRightVertex x (vertices ys zs) == elem x zs
```

snowleopard commented on src/Algebra/Graph/Bipartite/AdjacencyMap.hs:
```diff
@@ -0,0 +1,668 @@
+-- toBipartite (Algebra.Graph.AdjacencyMap.'Algebra.Graph.AdjacencyMap.edge' (Left 1) (Right "a")) == 'edge' 1 "a"
+-- @
+toBipartite :: (Ord a, Ord b) => AM.AdjacencyMap (Either a b) -> AdjacencyMap a b
+toBipartite m = BAM (Map.fromAscList [ (u, setRights vs) | (Left  u, vs) <- Map.toAscList (AM.adjacencyMap $ AM.symmetricClosure m)])
```
Please shorten the lines to simplify reviewing.

snowleopard commented on src/Algebra/Graph/Bipartite/AdjacencyMap.hs:
```diff
@@ -0,0 +1,668 @@
+-- fromBipartite ('edge' 1 2)     == 'Algebra.Graph.AdjacencyMap.edges' [(Left 1, Right 2), (Right 2, Left 1)]
+-- @
+fromBipartite :: (Ord a, Ord b) => AdjacencyMap a b -> AM.AdjacencyMap (Either a b)
+fromBipartite (BAM lr rl) = AM.overlays $
```
I think you could use `fromAdjacencySets` for faster implementation:

https://hackage.haskell.org/package/algebraic-graphs-0.4/docs/Algebra-Graph-AdjacencyMap.html#v:fromAdjacencySets

snowleopard commented on test/Algebra/Graph/Test/Bipartite/AdjacencyMap.hs:
```diff
@@ -0,0 +1,620 @@
+
+    putStrLn "\n============ Bipartite.AdjacencyMap.hasEdge ============"
+    test "hasEdge x y (fromGraph empty)                                 == False" $ \x y ->
+        not $ hasEdge (x :: Int) (y :: Int) $ fromGraph G.empty
```
It's typically a bit easier to read if these annotations are on the lambda, i.e. `\(x :: Int) (y :: Int) ->...`.

snowleopard commented on src/Algebra/Graph/Bipartite/AdjacencyMap.hs:
```diff
@@ -0,0 +1,668 @@
+show (1 + 2)                       == "vertices [] [1,2]"
+show ('swap' (1 + 2))                == "vertices [1,2] []"
+show (('swap' 1) * 2)                == "edge 1 2"
+show (('swap' 1) * 2 * ('swap' 3))     == "edges [(1,2),(3,2)]"
```
Note that functions bind more tightly than operators, so some of the above parentheses are redundant (also in tests).

adithyaov commented on src/Algebra/Graph/Bipartite/AdjacencyMap.hs:
```diff
@@ -0,0 +1,667 @@
+import qualified Data.Tuple
+
+{-| The 'Bipartite.AdjacencyMap' data type represents an __undirected__
+bipartite graph by two maps of vertices into their adjacency sets. The two type
```
I was told by @snowleopard  that it's best to leave out the implementation details from the documentation as the implementation might change later.

snowleopard replied:
Good point!

Note that it's not only about the implementation that might change in future, but also about unnecessary information in the documentation, which takes up the time of the reader.

vasalf replied:
Thanks for the comment.

I added this in order to show some difference between `Algebra.Graph.Bipartite.AdjacencyMap` and `Algebra.Graph.Bipartite` (if added in future). Now I see that there is no such difference in the documentation for `Algebra.Graph.AdjacencyMap`, so I removed it.

snowleopard commented:
@vasalf There is a test failure:

```
============ Eq (Bipartite.AdjacencyMap a b) ============
Test failure:
    (x == y) == ((leftAdjacencyMap x == rightAdjacencyMap x) && (leftAdjacencyMap y == rightAdjacencyMap y))
*** Failed! Falsified (after 3 tests):
empty
vertices [-2] [-2]
```

The full log is here: https://travis-ci.org/snowleopard/alga/jobs/550349421

Please investigate.

vasalf commented:
> @vasalf There is a test failure:
> 
> ```
> ============ Eq (Bipartite.AdjacencyMap a b) ============
> Test failure:
>     (x == y) == ((leftAdjacencyMap x == rightAdjacencyMap x) && (leftAdjacencyMap y == rightAdjacencyMap y))
> *** Failed! Falsified (after 3 tests):
> empty
> vertices [-2] [-2]
> ```
> 
> The full log is here: https://travis-ci.org/snowleopard/alga/jobs/550349421
> 
> Please investigate.

@snowleopard Thanks for pointing it out! It turned out that the test was wrong. I fixed it. Hope it will pass this time (it passes on my machine, but we know that means nothing). 

I also fixed a small bug in `Show` instance.

snowleopard commented:
Thanks, merged!
